src/go/pt-k8s-pxc-recovery/recover/recover.go

```diff
@@ -0,0 +1,285 @@
+	if err != nil {
+		return err
+	}
+	failedNodes := 0
+	for _, pod := range podNames {
+		logs, err := kubectl.RunCmd(c.Namespace, "logs", pod)
+		if err != nil {
+			return errors.New("error confirming crashed cluster status" + err.Error())
+		}
+		if strings.Contains(string(logs), "grastate.dat") && strings.Contains(string(logs), "safe_to_bootstrap") {
+			failedNodes++
+		}
+	}
+	if len(podNames) != failedNodes {
+		return errors.New("found more than one pod running, can't use recovery tool, please manually restart failed pods")
+	}
+	return nil
+}
+
+func (c *Cluster) PatchClusterImage(image string) error {
+	args := []string{
+		"patch",
+		"pxc",
+		c.Name,
+		"--type=merge",
+		`--patch={"spec":{"pxc":{"image":"` + image + `"}}}`,
+	}
+	_, err := kubectl.RunCmd(c.Namespace, args...)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Cluster) RestartPods() error {
+	for i := 0; i < c.Size; i++ {
+		args := []string{
+			"delete",
+			"pod",
+			c.Name + "-pxc-" + strconv.Itoa(i),
+			"--force",
+			"--grace-period=0",
+		}
+		_, err := kubectl.RunCmd(c.Namespace, args...)
+		if err != nil && !strings.ContainsAny(err.Error(), "pods") && !strings.ContainsAny(err.Error(), "not found") {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Cluster) CheckPodStatus(podID int) (bool, error) {
+	args := []string{
+		"get",
+		"pod", c.Name + "-pxc-" + strconv.Itoa(podID),
+		"-o",
+		"jsonpath='{.status.containerStatuses[0].ready}'",
+	}
+	output, err := kubectl.RunCmd(c.Namespace, args...)
+	if err != nil {
+		return false, err
+	}
+	if strings.Trim(output, "'") == "true" {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *Cluster) PodZeroReady() error {
+	podZeroStatus := false
+	var err error
+	for !podZeroStatus {
+		time.Sleep(time.Second * 10)
+		podZeroStatus, err = c.CheckPodStatus(0)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Cluster) CheckPodPhase(podID int, phase string) (bool, error) {
+	args := []string{
+		"get",
+		"pod", c.Name + "-pxc-" + strconv.Itoa(podID),
+		"-o",
+		"jsonpath='{.status.phase}'",
+	}
+	output, err := kubectl.RunCmd(c.Namespace, args...)
+	if err != nil {
+		return false, err
+	}
+	return strings.Trim(output, "'") == phase, nil
+}
+
+func (c *Cluster) AllPodsRunning() error {
+	for i := 0; i < c.Size; i++ {
+		running := false
+		for !running {
+			time.Sleep(time.Second * 10)
+			running, _ = c.CheckPodPhase(i, "Running")
+		}
+	}
+	return nil
+}
+
+func (c *Cluster) RunCommandInPod(podID int, cmd ...string) (string, error) {
+	args := []string{
+		"exec",
+		c.Name + "-pxc-" + strconv.Itoa(podID),
+		"--",
+	}
+	args = append(args, cmd...)
+	output, err := kubectl.RunCmd(c.Namespace, args...)
+	if err != nil {
+		return "", err
+	}
+	return output, nil
+}
+
+func (c *Cluster) SetSSTInProgress() error {
+	for i := 0; i < c.Size; i++ {
+		_, err := c.RunCommandInPod(i, "touch", "/var/lib/mysql/sst_in_progress")
+		if err != nil {
+			return err
```

Comment:
provide more context for the error. e.g. it would be nice to know on what pod error happened
the same in many places with errors

Reply:
these errors are kubectl error messages they are pretty verbose

Reply:
1. You're running a kubectl command via exec.Cmd. So error could happen out of the scope of kubectl
2. Some extra context won't make any worse. This tool is intended to be run by customers. If something went wrong we should have as much info as we can to tackle the problem. And we won't have a chance to adjust error messaging in the tool and ask to rerun it so we could understand e.g. on what pod error happened. 

And this is related not only to this code but to error handling in the tool in general. Wrap errors with some context info before return!!!